File: typescript/src/evm.ts
/**
 * HyperEVM Client — Ethereum JSON-RPC for Hyperliquid's EVM.
 *
 * Standard Ethereum JSON-RPC methods plus debug/trace capabilities.
 *
 * Example:
 *     import { EVM } from 'hyperliquid-sdk';
 *     const evm = new EVM("https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN");
 *     console.log(await evm.blockNumber());
 *     console.log(await evm.getBalance("0x..."));
 */

import { HyperliquidError } from './errors';

export interface EVMOptions {
  /** Use nanoreth path for debug/trace APIs (mainnet only) */
  debug?: boolean;
  /** Request timeout in milliseconds (default: 30000) */
  timeout?: number;
}

/**
 * HyperEVM Client — Ethereum JSON-RPC for Hyperliquid's EVM.
 *
 * Standard Ethereum methods plus debug/trace APIs (nanoreth path).
 *
 * Examples:
 *     const evm = new EVM("https://your-endpoint.hype-mainnet.quiknode.pro/TOKEN");
 *
 *     // Standard methods
 *     await evm.blockNumber();
 *     await evm.chainId();
 *     await evm.getBalance("0x...");
 *     await evm.getBlockByNumber(12345);
 *
 *     // Debug/Trace (mainnet only)
 *     const evmDebug = new EVM(endpoint, { debug: true });
 *     await evmDebug.traceTransaction("0x...");
 */
export class EVM {
  private readonly _baseUrl: string;
  private readonly _timeout: number;
  private _requestId: number = 0;

  constructor(endpoint: string, options: EVMOptions = {}) {
    this._baseUrl = this._buildUrl(endpoint, options.debug ?? false);
    this._timeout = options.timeout ?? 30000;
  }

  private _buildUrl(url: string, useNanoreth: boolean): string {
    const parsed = new URL(url);
    const base = `${parsed.protocol}//${parsed.host}`;
    const pathParts = parsed.pathname.split('/').filter((p) => p.length > 0);

    const knownPaths = new Set(['info', 'hypercore', 'evm', 'nanoreth']);
    let token: string | null = null;
    for (const part of pathParts) {
      if (!knownPaths.has(part)) {
        token = part;
        break;
      }
    }

    const path = useNanoreth ? 'nanoreth' : 'evm';
    if (token) {
      return `${base}/${token}/${path}`;
    }
    return `${base}/${path}`;
  }

  /** Make a JSON-RPC call. */
  private async _rpc<T = unknown>(
    method: string,
    params: unknown[] = []
  ): Promise<T> {
    this._requestId += 1;
    const payload = {
      jsonrpc: '2.0',
      method,
      params,
      id: this._requestId,
    };

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this._timeout);

    try {
      const response = await fetch(this._baseUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        throw new HyperliquidError(`Request failed: ${response.status}`, {
          code: 'HTTP_ERROR',
          raw: { body: await response.text() },
        });
      }

      const data = await response.json() as Record<string, unknown>;

      if (data.error) {
        const errorObj = data.error as Record<string, unknown>;
        throw new HyperliquidError(
          (errorObj.message as string) ?? 'RPC error',
          {
            code: String(errorObj.code ?? 'RPC_ERROR'),
            raw: data,
          }
        );
      }

      return data.result as T;
    } catch (error) {
      clearTimeout(timeoutId);

      if (error instanceof HyperliquidError) throw error;

      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          throw new HyperliquidError(`Request timed out after ${this._timeout}ms`, {
            code: 'TIMEOUT',
            raw: { method, timeout: this._timeout },
          });
        }
        throw new HyperliquidError(`Connection failed: ${error.message}`, {
          code: 'CONNECTION_ERROR',
          raw: { method, error: error.message },
        });
      }
      throw error;
    }
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // STANDARD ETHEREUM METHODS
  // ═══════════════════════════════════════════════════════════════════════════

  /** Get the current block number. */
  async blockNumber(): Promise<number> {
    const result = await this._rpc<string>('eth_blockNumber');
    return parseInt(result, 16);
  }

  /** Get the chain ID (999 mainnet, 998 testnet). */
  async chainId(): Promise<number> {
    const result = await this._rpc<string>('eth_chainId');
    return parseInt(result, 16);
  }

  /** Get the current gas price in wei. */
  async gasPrice(): Promise<bigint> {
    const result = await this._rpc<string>('eth_gasPrice');
    return BigInt(result);
  }

  /** Get account balance in wei. */
  async getBalance(address: string, block: number | string = 'latest'): Promise<bigint> {
    const blockParam = typeof block === 'number' ? `0x${block.toString(16)}` : block;
    const result = await this._rpc<string>('eth_getBalance', [address, blockParam]);
    return BigInt(result);
  }

  /** Get the nonce (transaction count) for an address. */
  async getTransactionCount(
    address: string,
    block: number | string = 'latest'
  ): Promise<number> {
    const blockParam = typeof block === 'number' ? `0x${block.toString(16)}` : block;
    const result = await this._rpc<string>('eth_getTransactionCount', [address, blockParam]);
    return parseInt(result, 16);
  }

  /** Get the contract bytecode at an address. */
  async getCode(address: string, block: number | string = 'latest'): Promise<string> {
    const blockParam = typeof block === 'number' ? `0x${block.toString(16)}` : block;
    return this._rpc('eth_getCode', [address, blockParam]);
  }

  /** Get storage at a specific position. */
  async getStorageAt(
    address: string,
    position: string,
    block: number | string = 'latest'
  ): Promise<string> {
    const blockParam = typeof block === 'number' ? `0x${block.toString(16)}` : block;
    return this._rpc('eth_getStorageAt', [address, position, blockParam]);
  }

  /** Execute a read-only call. */
  async call(
    tx: Record<string, unknown>,
    block: number | string = 'latest'
  ): Promise<string> {
    const blockParam = typeof block === 'number' ? `0x${block.toString(16)}` : block;
    return this._rpc('eth_call', [tx, blockParam]);
  }

  /** Estimate gas for a transaction. */
  async estimateGas(tx: Record<string, unknown>): Promise<bigint> {
    const result = await this._rpc<string>('eth_estimateGas', [tx]);
    return BigInt(result);
  }

  /** Submit a signed transaction. */
  async sendRawTransaction(signedTx: string): Promise<string> {
    return this._rpc('eth_sendRawTransaction', [signedTx]);
  }

  /** Get transaction by hash. */
  async getTransactionByHash(
    txHash: string
  ): Promise<Record<string, unknown> | null> {
    return this._rpc('eth_getTransactionByHash', [txHash]);
  }

  /** Get transaction receipt. */
  async getTransactionReceipt(
    txHash: string
  ): Promise<Record<string, unknown> | null> {
    return this._rpc('eth_getTransactionReceipt', [txHash]);
  }

  /** Get block by number. */
  async getBlockByNumber(
    blockNumber: number | string,
    fullTransactions: boolean = false
  ): Promise<Record<string, unknown> | null> {
    const blockParam =
      typeof blockNumber === 'number' ? `0x${blockNumber.toString(16)}` : blockNumber;
    return this._rpc('eth_getBlockByNumber', [blockParam, fullTransactions]);
  }

  /** Get block by hash. */
  async getBlockByHash(
    blockHash: string,
    fullTransactions: boolean = false
  ): Promise<Record<string, unknown> | null> {
    return this._rpc('eth_getBlockByHash', [blockHash, fullTransactions]);
  }

  /** Get logs matching filter (max 4 topics, 50 block range). */
  async getLogs(
    filterParams: Record<string, unknown>
  ): Promise<Array<Record<string, unknown>>> {
    return this._rpc('eth_getLogs', [filterParams]);
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // DEBUG/TRACE METHODS (requires debug=true, mainnet only)
  // ═══════════════════════════════════════════════════════════════════════════

  /** Trace a transaction's execution. */
  async debugTraceTransaction(
    txHash: string,
    tracerConfig?: Record<string, unknown>
  ): Promise<Record<string, unknown>> {
    const params: unknown[] = [txHash];
    if (tracerConfig) {
      params.push(tracerConfig);
    }
    return this._rpc('debug_traceTransaction', params);
  }

  /** Get trace of a transaction. */
  async traceTransaction(
    txHash: string
  ): Promise<Array<Record<string, unknown>>> {
    return this._rpc('trace_transaction', [txHash]);
  }

  /** Get traces of all transactions in a block. */
  async traceBlock(
    blockNumber: number | string
  ): Promise<Array<Record<string, unknown>>> {
    const blockParam =
      typeof blockNumber === 'number' ? `0x${blockNumber.toString(16)}` : blockNumber;
    return this._rpc('trace_block', [blockParam]);
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // ADDITIONAL STANDARD METHODS
  // ═══════════════════════════════════════════════════════════════════════════

  /** Get the network version. */
  async netVersion(): Promise<string> {
    return this._rpc('net_version');
  }

  /** Get the client version. */
  async web3ClientVersion(): Promise<string> {
    return this._rpc('web3_clientVersion');
  }

  /** Check if the node is syncing. */
  async syncing(): Promise<boolean | Record<string, unknown>> {
    return this._rpc('eth_syncing');
  }

  /** Get list of accounts (usually empty for remote nodes). */
  async accounts(): Promise<string[]> {
    return this._rpc('eth_accounts');
  }

  /** Get fee history for a range of blocks. */
  async feeHistory(
    blockCount: number,
    newestBlock: number | string,
    rewardPercentiles?: number[]
  ): Promise<Record<string, unknown>> {
    const blockParam =
      typeof newestBlock === 'number' ? `0x${newestBlock.toString(16)}` : newestBlock;
    const params: unknown[] = [`0x${blockCount.toString(16)}`, blockParam];
    if (rewardPercentiles) {
      params.push(rewardPercentiles);
    }
    return this._rpc('eth_feeHistory', params);
  }

  /** Get max priority fee per gas. */
  async maxPriorityFeePerGas(): Promise<bigint> {
    const result = await this._rpc<string>('eth_maxPriorityFeePerGas');
    return BigInt(result);
  }

  /** Get all receipts for a block. */
  async getBlockReceipts(
    blockNumber: number | string
  ): Promise<Array<Record<string, unknown>>> {
    const blockParam =
      typeof blockNumber === 'number' ? `0x${blockNumber.toString(16)}` : blockNumber;
    return this._rpc('eth_getBlockReceipts', [blockParam]);
  }

  /** Get transaction count in a block by hash. */
  async getBlockTransactionCountByHash(blockHash: string): Promise<number> {
    const result = await this._rpc<string>('eth_getBlockTransactionCountByHash', [
      blockHash,
    ]);
    return parseInt(result, 16);
  }

  /** Get transaction count in a block by number. */
  async getBlockTransactionCountByNumber(
    blockNumber: number | string
  ): Promise<number> {
    const blockParam =
      typeof blockNumber === 'number' ? `0x${blockNumber.toString(16)}` : blockNumber;
    const result = await this._rpc<string>('eth_getBlockTransactionCountByNumber', [
      blockParam,
    ]);
    return parseInt(result, 16);
  }

  /** Get transaction by block hash and index. */
  async getTransactionByBlockHashAndIndex(
    blockHash: string,
    index: number
  ): Promise<Record<string, unknown> | null> {
    return this._rpc('eth_getTransactionByBlockHashAndIndex', [
      blockHash,
      `0x${index.toString(16)}`,
    ]);
  }

  /** Get transaction by block number and index. */
  async getTransactionByBlockNumberAndIndex(
    blockNumber: number | string,
    index: number
  ): Promise<Record<string, unknown> | null> {
    const blockParam =
      typeof blockNumber === 'number' ? `0x${blockNumber.toString(16)}` : blockNumber;
    return this._rpc('eth_getTransactionByBlockNumberAndIndex', [
      blockParam,
      `0x${index.toString(16)}`,
    ]);
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // HYPERLIQUID-SPECIFIC EVM METHODS
  // ═══════════════════════════════════════════════════════════════════════════

  /** Get gas price for big blocks. */
  async bigBlockGasPrice(): Promise<bigint> {
    const result = await this._rpc<string>('eth_bigBlockGasPrice');
    return BigInt(result);
  }

  /** Check if using big blocks. */
  async usingBigBlocks(): Promise<boolean> {
    return this._rpc('eth_usingBigBlocks');
  }

  /** Get system transactions by block hash. */
  async getSystemTxsByBlockHash(
    blockHash: string
  ): Promise<Array<Record<string, unknown>>> {
    return this._rpc('eth_getSystemTxsByBlockHash', [blockHash]);
  }

  /** Get system transactions by block number. */
  async getSystemTxsByBlockNumber(
    blockNumber: number | string
  ): Promise<Array<Record<string, unknown>>> {
    const blockParam =
      typeof blockNumber === 'number' ? `0x${blockNumber.toString(16)}` : blockNumber;
    return this._rpc('eth_getSystemTxsByBlockNumber', [blockParam]);
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // ADDITIONAL DEBUG METHODS
  // ═══════════════════════════════════════════════════════════════════════════

  /** Get bad blocks. */
  async debugGetBadBlocks(): Promise<Array<Record<string, unknown>>> {
    return this._rpc('debug_getBadBlocks');
  }

  /** Get raw block data. */
  async debugGetRawBlock(blockNumber: number | string): Promise<string> {
    const blockParam =
      typeof blockNumber === 'number' ? `0x${blockNumber.toString(16)}` : blockNumber;
    return this._rpc('debug_getRawBlock', [blockParam]);
  }

  /** Get raw block header. */
  async debugGetRawHeader(blockNumber: number | string): Promise<string> {
    const blockParam =
      typeof blockNumber === 'number' ? `0x${blockNumber.toString(16)}` : blockNumber;
    return this._rpc('debug_getRawHeader', [blockParam]);
  }

  /** Get raw receipts for a block. */
  async debugGetRawReceipts(blockNumber: number | string): Promise<string[]> {
    const blockParam =
      typeof blockNumber === 'number' ? `0x${blockNumber.toString(16)}` : blockNumber;
    return this._rpc('debug_getRawReceipts', [blockParam]);
  }

  /** Get raw transaction data. */
  async debugGetRawTransaction(txHash: string): Promise<string> {
    return this._rpc('debug_getRawTransaction', [txHash]);
  }

  /** Get storage range at a specific point. */
  async debugStorageRangeAt(
    blockHash: string,
    txIndex: number,
    contractAddress: string,
    keyStart: string,
    maxResult: number
  ): Promise<Record<string, unknown>> {
    return this._rpc('debug_storageRangeAt', [
      blockHash,
      txIndex,
      contractAddress,
      keyStart,
      maxResult,
    ]);
  }

  /** Trace a block by RLP. */
  async debugTraceBlock(
    blockRlp: string,
    tracerConfig?: Record<string, unknown>
  ): Promise<Array<Record<string, unknown>>> {
    const params: unknown[] = [blockRlp];
    if (tracerConfig) {
      params.push(tracerConfig);
    }
    return this._rpc('debug_traceBlock', params);
  }

  /** Trace a block by hash. */
  async debugTraceBlockByHash(
    blockHash: string,
    tracerConfig?: Record<string, unknown>
  ): Promise<Array<Record<string, unknown>>> {
    const params: unknown[] = [blockHash];
    if (tracerConfig) {
      params.push(tracerConfig);
    }
    return this._rpc('debug_traceBlockByHash', params);
  }

  /** Trace a block by number. */
  async debugTraceBlockByNumber(
    blockNumber: number | string,
    tracerConfig?: Record<string, unknown>
  ): Promise<Array<Record<string, unknown>>> {
    const blockParam =
      typeof blockNumber === 'number' ? `0x${blockNumber.toString(16)}` : blockNumber;
    const params: unknown[] = [blockParam];
    if (tracerConfig) {
      params.push(tracerConfig);
    }
    return this._rpc('debug_traceBlockByNumber', params);
  }

  /** Trace a call. */
  async debugTraceCall(
    tx: Record<string, unknown>,
    block: number | string = 'latest',
    tracerConfig?: Record<string, unknown>
  ): Promise<Record<string, unknown>> {
    const blockParam = typeof block === 'number' ? `0x${block.toString(16)}` : block;
    const params: unknown[] = [tx, blockParam];
    if (tracerConfig) {
      params.push(tracerConfig);
    }
    return this._rpc('debug_traceCall', params);
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // ADDITIONAL TRACE METHODS
  // ═══════════════════════════════════════════════════════════════════════════

  /** Trace a call with specified trace types. */
  async traceCall(
    tx: Record<string, unknown>,
    traceTypes: string[],
    block: number | string = 'latest'
  ): Promise<Record<string, unknown>> {
    const blockParam = typeof block === 'number' ? `0x${block.toString(16)}` : block;
    return this._rpc('trace_call', [tx, traceTypes, blockParam]);
  }

  /** Trace multiple calls. */
  async traceCallMany(
    calls: Array<[Record<string, unknown>, string[]]>,
    block: number | string = 'latest'
  ): Promise<Array<Record<string, unknown>>> {
    const blockParam = typeof block === 'number' ? `0x${block.toString(16)}` : block;
    return this._rpc('trace_callMany', [calls, blockParam]);
  }

  /** Filter traces. */
  async traceFilter(
    filterParams: Record<string, unknown>
  ): Promise<Array<Record<string, unknown>>> {
    return this._rpc('trace_filter', [filterParams]);
  }

  /** Trace a raw transaction. */
  async traceRawTransaction(
    rawTx: string,
    traceTypes: string[]
  ): Promise<Record<string, unknown>> {
    return this._rpc('trace_rawTransaction', [rawTx, traceTypes]);
  }

  /** Replay and trace all transactions in a block. */
  async traceReplayBlockTransactions(
    blockNumber: number | string,
    traceTypes: string[]
  ): Promise<Array<Record<string, unknown>>> {
    const blockParam =
      typeof blockNumber === 'number' ? `0x${blockNumber.toString(16)}` : blockNumber;
    return this._rpc('trace_replayBlockTransactions', [blockParam, traceTypes]);
  }

  /** Replay and trace a transaction. */
  async traceReplayTransaction(
    txHash: string,
    traceTypes: string[]
  ): Promise<Record<string, unknown>> {
    return this._rpc('trace_replayTransaction', [txHash, traceTypes]);
  }
}

File: rust/src/evm.rs
//! EVM (HyperEVM) client for Hyperliquid.
//!
//! Provides Ethereum JSON-RPC compatibility for the Hyperliquid EVM.

use serde_json::{json, Value};
use std::sync::Arc;

use crate::client::HyperliquidSDKInner;
use crate::error::Result;

/// EVM API client (Ethereum JSON-RPC)
pub struct EVM {
    inner: Arc<HyperliquidSDKInner>,
    debug: bool,
}

impl EVM {
    pub(crate) fn new(inner: Arc<HyperliquidSDKInner>) -> Self {
        Self {
            inner,
            debug: false,
        }
    }

    /// Enable debug/trace methods (mainnet only)
    pub fn with_debug(mut self, debug: bool) -> Self {
        self.debug = debug;
        self
    }

    /// Get the EVM endpoint URL
    fn evm_url(&self) -> String {
        self.inner.evm_url(self.debug)
    }

    /// Make a JSON-RPC request
    async fn rpc(&self, method: &str, params: Value) -> Result<Value> {
        let url = self.evm_url();

        let body = json!({
            "jsonrpc": "2.0",
            "method": method,
            "params": params,
            "id": 1,
        });

        let response = self
            .inner
            .http_client
            .post(&url)
            .json(&body)
            .send()
            .await?;

        let status = response.status();
        let text = response.text().await?;

        if !status.is_success() {
            return Err(crate::Error::NetworkError(format!(
                "EVM request failed {}: {}",
                status, text
            )));
        }

        let result: Value = serde_json::from_str(&text)?;

        // Check for JSON-RPC error
        if let Some(error) = result.get("error") {
            let message = error
                .get("message")
                .and_then(|m| m.as_str())
                .unwrap_or("Unknown error");
            return Err(crate::Error::ApiError {
                code: crate::error::ErrorCode::Unknown,
                message: message.to_string(),
                guidance: "Check your EVM request parameters.".to_string(),
                raw: Some(error.to_string()),
            });
        }

        Ok(result.get("result").cloned().unwrap_or(Value::Null))
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Chain Info
    // ──────────────────────────────────────────────────────────────────────────

    /// Get the current block number
    pub async fn block_number(&self) -> Result<u64> {
        let result = self.rpc("eth_blockNumber", json!([])).await?;
        parse_hex_u64(&result)
    }

    /// Get the chain ID
    pub async fn chain_id(&self) -> Result<u64> {
        let result = self.rpc("eth_chainId", json!([])).await?;
        parse_hex_u64(&result)
    }

    /// Check if the node is syncing
    pub async fn syncing(&self) -> Result<Value> {
        self.rpc("eth_syncing", json!([])).await
    }

    /// Get current gas price
    pub async fn gas_price(&self) -> Result<u64> {
        let result = self.rpc("eth_gasPrice", json!([])).await?;
        parse_hex_u64(&result)
    }

    /// Get network version
    pub async fn net_version(&self) -> Result<String> {
        let result = self.rpc("net_version", json!([])).await?;
        Ok(result.as_str().unwrap_or("").to_string())
    }

    /// Get client version
    pub async fn web3_client_version(&self) -> Result<String> {
        let result = self.rpc("web3_clientVersion", json!([])).await?;
        Ok(result.as_str().unwrap_or("").to_string())
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Account
    // ──────────────────────────────────────────────────────────────────────────

    /// Get balance of an address
    pub async fn get_balance(&self, address: &str, block: Option<&str>) -> Result<String> {
        let block = block.unwrap_or("latest");
        let result = self.rpc("eth_getBalance", json!([address, block])).await?;
        Ok(result.as_str().unwrap_or("0x0").to_string())
    }

    /// Get transaction count (nonce) of an address
    pub async fn get_transaction_count(&self, address: &str, block: Option<&str>) -> Result<u64> {
        let block = block.unwrap_or("latest");
        let result = self
            .rpc("eth_getTransactionCount", json!([address, block]))
            .await?;
        parse_hex_u64(&result)
    }

    /// Get code at an address
    pub async fn get_code(&self, address: &str, block: Option<&str>) -> Result<String> {
        let block = block.unwrap_or("latest");
        let result = self.rpc("eth_getCode", json!([address, block])).await?;
        Ok(result.as_str().unwrap_or("0x").to_string())
    }

    /// Get storage at a position
    pub async fn get_storage_at(
        &self,
        address: &str,
        position: &str,
        block: Option<&str>,
    ) -> Result<String> {
        let block = block.unwrap_or("latest");
        let result = self
            .rpc("eth_getStorageAt", json!([address, position, block]))
            .await?;
        Ok(result.as_str().unwrap_or("0x0").to_string())
    }

    /// Get list of accounts (usually empty for remote nodes)
    pub async fn accounts(&self) -> Result<Vec<String>> {
        let result = self.rpc("eth_accounts", json!([])).await?;
        Ok(result
            .as_array()
            .map(|arr| {
                arr.iter()
                    .filter_map(|v| v.as_str().map(|s| s.to_string()))
                    .collect()
            })
            .unwrap_or_default())
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Transactions
    // ──────────────────────────────────────────────────────────────────────────

    /// Call a contract (read-only)
    pub async fn call(&self, tx: &Value, block: Option<&str>) -> Result<String> {
        let block = block.unwrap_or("latest");
        let result = self.rpc("eth_call", json!([tx, block])).await?;
        Ok(result.as_str().unwrap_or("0x").to_string())
    }

    /// Estimate gas for a transaction
    pub async fn estimate_gas(&self, tx: &Value) -> Result<u64> {
        let result = self.rpc("eth_estimateGas", json!([tx])).await?;
        parse_hex_u64(&result)
    }

    /// Send a raw (signed) transaction
    pub async fn send_raw_transaction(&self, signed_tx: &str) -> Result<String> {
        let result = self.rpc("eth_sendRawTransaction", json!([signed_tx])).await?;
        Ok(result.as_str().unwrap_or("").to_string())
    }

    /// Get transaction by hash
    pub async fn get_transaction_by_hash(&self, tx_hash: &str) -> Result<Value> {
        self.rpc("eth_getTransactionByHash", json!([tx_hash])).await
    }

    /// Get transaction receipt
    pub async fn get_transaction_receipt(&self, tx_hash: &str) -> Result<Value> {
        self.rpc("eth_getTransactionReceipt", json!([tx_hash])).await
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Blocks
    // ──────────────────────────────────────────────────────────────────────────

    /// Get block by number
    pub async fn get_block_by_number(
        &self,
        block_number: &str,
        full_transactions: bool,
    ) -> Result<Value> {
        self.rpc(
            "eth_getBlockByNumber",
            json!([block_number, full_transactions]),
        )
        .await
    }

    /// Get block by hash
    pub async fn get_block_by_hash(&self, block_hash: &str, full_transactions: bool) -> Result<Value> {
        self.rpc(
            "eth_getBlockByHash",
            json!([block_hash, full_transactions]),
        )
        .await
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Logs
    // ──────────────────────────────────────────────────────────────────────────

    /// Get logs matching a filter
    pub async fn get_logs(&self, filter: &Value) -> Result<Value> {
        self.rpc("eth_getLogs", json!([filter])).await
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Fees
    // ──────────────────────────────────────────────────────────────────────────

    /// Get fee history
    pub async fn fee_history(
        &self,
        block_count: u64,
        newest_block: &str,
        reward_percentiles: Option<&[f64]>,
    ) -> Result<Value> {
        let percentiles = reward_percentiles.unwrap_or(&[]);
        self.rpc(
            "eth_feeHistory",
            json!([format!("0x{:x}", block_count), newest_block, percentiles]),
        )
        .await
    }

    /// Get max priority fee per gas
    pub async fn max_priority_fee_per_gas(&self) -> Result<u64> {
        let result = self.rpc("eth_maxPriorityFeePerGas", json!([])).await?;
        parse_hex_u64(&result)
    }

    /// Get block receipts
    pub async fn get_block_receipts(&self, block_number: &str) -> Result<Value> {
        self.rpc("eth_getBlockReceipts", json!([block_number])).await
    }

    /// Get block transaction count by hash
    pub async fn get_block_transaction_count_by_hash(&self, block_hash: &str) -> Result<u64> {
        let result = self
            .rpc("eth_getBlockTransactionCountByHash", json!([block_hash]))
            .await?;
        parse_hex_u64(&result)
    }

    /// Get block transaction count by number
    pub async fn get_block_transaction_count_by_number(&self, block_number: &str) -> Result<u64> {
        let result = self
            .rpc("eth_getBlockTransactionCountByNumber", json!([block_number]))
            .await?;
        parse_hex_u64(&result)
    }

    /// Get transaction by block hash and index
    pub async fn get_transaction_by_block_hash_and_index(
        &self,
        block_hash: &str,
        index: u64,
    ) -> Result<Value> {
        self.rpc(
            "eth_getTransactionByBlockHashAndIndex",
            json!([block_hash, format!("0x{:x}", index)]),
        )
        .await
    }

    /// Get transaction by block number and index
    pub async fn get_transaction_by_block_number_and_index(
        &self,
        block_number: &str,
        index: u64,
    ) -> Result<Value> {
        self.rpc(
            "eth_getTransactionByBlockNumberAndIndex",
            json!([block_number, format!("0x{:x}", index)]),
        )
        .await
    }

    // ──────────────────────────────────────────────────────────────────────────
    // Debug/Trace (requires debug=true, mainnet only)
    // ──────────────────────────────────────────────────────────────────────────

    /// Debug trace a transaction
    pub async fn debug_trace_transaction(
        &self,
        tx_hash: &str,
        tracer_config: Option<&Value>,
    ) -> Result<Value> {
        if !self.debug {
            return Err(crate::Error::ConfigError(
                "Debug mode not enabled. Use .with_debug(true)".to_string(),
            ));
        }
        let config = tracer_config.cloned().unwrap_or(json!({}));
        self.rpc("debug_traceTransaction", json!([tx_hash, config]))
            .await
    }

    /// Trace a transaction
    pub async fn trace_transaction(&self, tx_hash: &str) -> Result<Value> {
        if !self.debug {
            return Err(crate::Error::ConfigError(
                "Debug mode not enabled. Use .with_debug(true)".to_string(),
            ));
        }
        self.rpc("trace_transaction", json!([tx_hash])).await
    }

    /// Trace a block
    pub async fn trace_block(&self, block_number: &str) -> Result<Value> {
        if !self.debug {
            return Err(crate::Error::ConfigError(
                "Debug mode not enabled. Use .with_debug(true)".to_string(),
            ));
        }
        self.rpc("trace_block", json!([block_number])).await
    }
}

/// Parse a hex string to u64
fn parse_hex_u64(value: &Value) -> Result<u64> {
    let s = value.as_str().unwrap_or("0x0");
    let s = s.trim_start_matches("0x");
    u64::from_str_radix(s, 16)
        .map_err(|e| crate::Error::JsonError(format!("Invalid hex number: {}", e)))
}
